README.md
# Kui Nameplates PVP
Simple addon for Kui Nameplates. Download releases [at Curse](https://mods.curse.com/addons/wow/281232-kuinameplates-pvp).

Add faction or PVP icon for nameplates 


changelog
Update version for twitch client.


main.lua
local folder,ns = ...
local addon = KuiNameplates
local core = KuiNameplatesCore
local mod = addon:NewPlugin('PVP')
local opt, events = KuiPVPConfig, {}
local frame_name = 'KuiPVPConfig'

-- create ######################################################################
function mod:Show(f)
    if (opt.env.iconFaction or opt.env.iconCombat or opt.env.iconOwnFaction)  then
        f:UpdatePVPIcon(f)
    else
        f.pvpIcon:Hide()
    end
end


function mod:Create(f)
    self:CreatePVPIcon(f)
end

local function UpdatePVPIcon(f)
    if (opt.env.iconFaction) and (opt.env.iconOwnFaction) and (UnitFactionGroup(f.unit)) then
        if (UnitFactionGroup(f.unit) ~= UnitFactionGroup("player")) then
        f.pvpIcon:SetTexture("Interface\\TargetingFrame\\UI-PVP-"..UnitFactionGroup(f.unit));
        f.pvpIcon:SetTexCoord(0,0.62,0,0.62);
        f.pvpIcon:SetWidth(opt.env.iconSize);
        f.pvpIcon:SetHeight(opt.env.iconSize);
        f.pvpIcon:Show()
        else
            f.pvpIcon:Hide()
        end
    elseif (opt.env.iconFaction) and (UnitIsPVPFreeForAll(f.unit)) then
        f.pvpIcon:SetTexture("Interface\\TargetingFrame\\UI-PVP-FFA");
        f.pvpIcon:SetTexCoord(0,0.62,0,0.62);
        f.pvpIcon:SetWidth(opt.env.iconSize);
        f.pvpIcon:SetHeight(opt.env.iconSize);
        f.pvpIcon:Show();
    elseif (opt.env.iconFaction) and (UnitFactionGroup(f.unit)) then
        f.pvpIcon:SetTexture("Interface\\TargetingFrame\\UI-PVP-"..UnitFactionGroup(f.unit));
        f.pvpIcon:SetTexCoord(0,0.62,0,0.62);
        f.pvpIcon:SetWidth(opt.env.iconSize);
        f.pvpIcon:SetHeight(opt.env.iconSize);
        f.pvpIcon:Show()
    elseif (opt.env.iconCombat) and (UnitAffectingCombat(f.unit)) then
        f.pvpIcon:SetTexture("Interface\\CharacterFrame\\UI-StateIcon");
        f.pvpIcon:SetTexCoord(0.5,1,0,0.5);
        f.pvpIcon:SetWidth(opt.env.iconSize);
        f.pvpIcon:SetHeight(opt.env.iconSize);
        f.pvpIcon:Show();
    else
        f.pvpIcon:Hide()
    end
end

function mod:CreatePVPIcon(f)
    local pvpIcon = f:CreateTexture(nil,"BACKGROUND");
    pvpIcon:SetPoint(opt.MirrorAnchors["RIGHT"],f,"RIGHT");
    f.pvpIcon = pvpIcon
    f.UpdatePVPIcon = UpdatePVPIcon

end

-- initialise ##################################################################
function mod:Initialise()
    self:RegisterMessage('Show')
    self:RegisterMessage('Create')
end
-- Event Holders ###############################################################
function events:ADDON_LOADED(addon_name)
    if addon_name == 'Kui_Nameplates_PVP' then
        init()
        local version = opt:CreateFontString(nil, 'ARTWORK', 'GameFontHighlight')
        version:SetAlpha(.5)
        version:SetPoint('TOPRIGHT',-12,-10)
        version:SetText(string.format(
            opt.info.header,
            opt.info.name,
            opt.info.author,
            opt.info.version
        ))

        local icon = CreateCheckBox(opt, 'iconFaction')
        icon:SetPoint("TOP", -250, -65)
        local icon = CreateCheckBox(opt, 'iconCombat')
        icon:SetPoint("TOP", -250, -85)
        local icon = CreateCheckBox(opt, 'iconOwnFaction')
        icon:SetPoint("TOP", -250, -125)
        local icon = CreateCheckBox(opt, 'Debugger')
        icon:SetPoint("BOTTOMLEFT", 0, 0)
        local iconSize = CreateSlider(opt, 'iconSize', 24, 72)
        iconSize:SetPoint("TOP", -140, -170)


    end
end

function events:PLAYER_LOGOUT()
    KuiPVPSaved = opt.env
end

opt:SetScript('OnEvent', function(self, event, ...)
    events[event](self, ...)
end)
for k, v in pairs(events) do
    opt:RegisterEvent(k)
end

config.lua
local folder,ns = ...
local opt = CreateFrame('FRAME','KuiPVPConfig',InterfaceOptionsFramePanelContainer)
opt.name = 'Kui |cff9966ffPVP'
InterfaceOptions_AddCategory(opt)

-- locale ######################################################################
do
    local L = {}
    function opt:Locale(region)
        assert(type(region) == 'string')
        if region == 'enGB' or region == GetLocale() then
            return L
        end
    end
    function opt:GetLocale()
        return L
    end
end

opt.info = {
    name = 'KuiNameplates: PVP',
    version = '0.0.9',
    author = 'GEOS',
    header = '%s by %s, version %s'
}
opt.titles = {
    iconFaction = 'Show Icon Faction',
    iconCombat = 'Show Icon Combat',
    --iconAnchor = 'Icon Anchor',
    iconSize = 'Change Icon Size',
    iconOwnFaction = 'Show Icon Own Faction'
}

opt.env = {
    iconFaction = true,
    iconCombat = true,
    iconAnchor = "RIGHT",
    iconSize = 24,
    iconOwnFaction = false,
    Debugger = false,
}

opt.Anchor = {
    RIGHT = "RIGHT",
    LEFT = "LEFT",
    TOP = "TOP",
    BOTTOM = "BOTTOM",
    TOPLEFT = "TOPLEFT",
    TOPRIGHT = "TOPRIGHT",
    BOTTOMLEFT = "BOTTOMLEFT",
    BOTTOMRIGHT = "BOTTOMRIGHT",
    RIGHT = "LEFT",
    CENTER = "CENTER"
}

opt.MirrorAnchors = {
    TOP = "BOTTOM",
    TOPLEFT = "TOPRIGHT",
    TOPRIGHT = "TOPLEFT",
    BOTTOM = "TOP",
    BOTTOMLEFT = "BOTTOMRIGHT",
    BOTTOMRIGHT = "BOTTOMLEFT",
    LEFT = "RIGHT",
    RIGHT = "LEFT",
    CENTER = "CENTER",
};


helpers.lua
local folder,ns = ...
local addon = KuiNameplates
local core = KuiNameplatesCore
local opt = KuiPVPConfig
local frame_name = 'KuiPVPConfig'
local L = opt:GetLocale()

-- init ########################################################################
function init()
    if KuiPVPSaved == nil then
        KuiPVPSaved = opt.env
    else
        opt.env = KuiPVPSaved
    end
end
-- helper functions ############################################################
function RegisterDebug(...)
    if opt.env.Debugger then
        print('|cff9966ffKui PVP: |r', ...)
    end
end
function ResetFrames()
    for i, f in addon:Frames() do
        if f:IsShown() then
            local unit = f.unit
            f.handler:OnHide()
            f.handler:OnUnitAdded(unit)
        end
    end
end
function IsInTable(t, val)
    for i, v in pairs(t) do
        if v == val then return true end
    end
    return false
end
function IsHealer(unit)
    for i, id in pairs(opt.healers) do
        if id == unit then
            RegisterDebug(select(6, GetPlayerInfoByGUID(unit)), 'is a healer')
            return true
        end
    end
    return false
end
-- element helpers #############################################################
local function CheckBoxOnClick(self)
    opt.env[self:GetName()] = self:GetChecked()
    if self:GetChecked() then
        PlaySound(PlaySoundKitID and "igMainMenuOptionCheckBoxOn" or 856)
    else
        PlaySound(PlaySoundKitID and "igMainMenuOptionCheckBoxOff" or 857)
    end
    ResetFrames()
end
local function OnEnter(self)
    GameTooltip:SetOwner(self,'ANCHOR_TOPLEFT')
    GameTooltip:SetWidth(200)
    GameTooltip:AddLine(
        self.env and (L.titles[self.env] or self.env) or
                self.label and self.label:GetText()
    )

    if self.env and L.tooltips[self.env] then
        GameTooltip:AddLine(L.tooltips[self.env], 1,1,1,true)
    end

    GameTooltip:Show()
end
local function OnLeave(self)
    GameTooltip:Hide()
end
local function GenericOnShow(self)
    if self.enabled then
        if self.enabled(opt.env) then
            self:Enable()
        else
            self:Disable()
        end
    end
end

do
local function SliderOnChanged(self,v)
    -- copy value to display text
    if not v then
        v = self:GetValue()
    end
    -- round value for display to hide floating point errors
    local r_v = string.format('%.4f',v)
    r_v = string.gsub(r_v,'0+$','')
    r_v = string.gsub(r_v,'%.$','')
    self.display:SetText(r_v)
end
local function Get(self)
    if self.env and opt.env[self.env] then
        self:SetValue(opt.env[self.env])
        -- set text to correct value if outside min/max
        SliderOnChanged(self,opt.env[self.env])
    end
end
local function Set(self,v)
    if not self:IsEnabled() then return end
    if self.env and opt.env[self.env] then
        --opt.config:SetConfig(self.env,v or self:GetValue())
        opt.env[self.env] = v or self:GetValue()
    end
end
local function SliderOnShow(self)
    if not opt.env then return end
    self:Get()
    GenericOnShow(self)
end
local function SliderOnMouseUp(self)
    self:Set()
end
local function SliderOnMouseWheel(self,delta)
    if not self:IsEnabled() then return end
    if delta > 0 then
        delta = self:GetValueStep()
    else
        delta = -self:GetValueStep()
    end
    self:SetValue(self:GetValue()+delta)
    self:Set()
end
local function SliderSetMinMaxValues(self,min,max)
    self:orig_SetMinMaxValues(min,max)
    self.Low:SetText(min)
    self.High:SetText(max)
end
local function EditBox_OnFocusGained(self)
    self:HighlightText()
end
local function EditBox_OnEscapePressed(self)
    self:ClearFocus()
    self:HighlightText(0,0)

    -- revert to current value
    SliderOnShow(self:GetParent())
end
local function EditBox_OnEnterPressed(self)
    -- dumb-verify input
    local v = tonumber(self:GetText())

    if v then
        -- display change
        self:GetParent():SetValue(v)
        -- push to config
        self:GetParent():Set(v)
    else
        EditBox_OnEscapePressed(self)
    end

    -- re-grab focus
    self:SetFocus()
end
function CreateSlider(parent, name, min, max)
        RegisterDebug('KPVP Creating Slider:', name)
        local slider = CreateFrame('Slider',frame_name..name..'Slider',parent,'OptionsSliderTemplate')
        slider:SetWidth(190)
        slider:SetHeight(15)
        slider:SetOrientation('HORIZONTAL')
        slider:SetThumbTexture('interface/buttons/ui-sliderbar-button-horizontal')
        slider:SetObeyStepOnDrag(true)
        slider:EnableMouseWheel(true)

        -- TODO inc/dec buttons
        local label = slider:CreateFontString(slider:GetName()..'Label','ARTWORK','GameFontNormal')
        label:SetText(L.titles[name] or name or 'Slider')
        label:SetPoint('BOTTOM',slider,'TOP')

        local display = CreateFrame('EditBox',nil,slider)
        display:SetFontObject('GameFontHighlightSmall')
        display:SetSize(50,15)
        display:SetPoint('TOP',slider,'BOTTOM',0,1)
        display:SetJustifyH('CENTER')
        display:SetAutoFocus(false)
        display:SetMultiLine(true)
        display:SetBackdrop({
            bgFile='interface/buttons/white8x8',
            edgeFile='interface/buttons/white8x8',
            edgeSize=1
        })
        display:SetBackdropBorderColor(1,1,1,.2)
        display:SetBackdropColor(0,0,0,.5)

        display:SetScript('OnEditFocusGained',EditBox_OnFocusGained)
        display:SetScript('OnEditFocusLost',EditBox_OnEscapePressed)
        display:SetScript('OnEnterPressed',EditBox_OnEnterPressed)
        display:SetScript('OnEscapePressed',EditBox_OnEscapePressed)

        slider.orig_SetMinMaxValues = slider.SetMinMaxValues
        slider.SetMinMaxValues = SliderSetMinMaxValues

        slider.env = name
        slider.label = label
        slider.display = display

        slider:HookScript('OnEnter',OnEnter)
        slider:HookScript('OnLeave',OnLeave)
        slider:HookScript('OnShow',SliderOnShow)
        slider:HookScript('OnValueChanged',SliderOnChanged)
        slider:HookScript('OnMouseUp',SliderOnMouseUp)
        slider:HookScript('OnMouseWheel',SliderOnMouseWheel)

        slider.Get = Get
        slider.Set = Set

        slider:SetValueStep(1)
        slider:SetMinMaxValues(min or 0, max or 100)

        if name and type(parent.elements) == 'table' then
            parent.elements[name] = slider
        end
        return slider
        --[[
        local slider = CreateFrame("Slider", name, parent, "OptionsSliderTemplate")
        slider:SetWidth(190)
        slider:SetHeight(15)
        slider:SetOrientation('HORIZONTAL')
        slider:SetThumbTexture('interface/buttons/ui-sliderbar-button-horizontal')
        slider:SetObeyStepOnDrag(true)
        slider:EnableMouseWheel(true)
        slider:SetMinMaxValues(24, 72)
        slider:SetValueStep(1)

        slider:SetScript("OnValueChanged", SliderChangeValue)

        slider.label = parent:CreateFontString(nil, 'ARTWORK', 'GameFontHighlight')
        slider.label:SetText(opt.titles[name] or name or '!MissingTitle')
        slider.label:SetPoint('LEFT', slider, 'TOPLEFT',0,10)

        return slider
        --]]
    end
end
function CreateCheckBox(parent, name)
    RegisterDebug('KPVP Creating Check Box:', name)
    local check = CreateFrame('CheckButton', name, parent, 'OptionsBaseCheckButtonTemplate')

    check:SetScript('OnClick',CheckBoxOnClick)

    check.label = parent:CreateFontString(nil, 'ARTWORK', 'GameFontHighlight')
    check.label:SetText(L.titles[name] or name or '!MissingTitle')
    check.label:SetPoint('LEFT', check, 'RIGHT')

    check:SetChecked(opt.env[name])

    return check
end


